warehouse/accounts/views.py

```diff
@@ -122,38 +126,123 @@ def login(request, redirect_field_name=REDIRECT_FIELD_NAME, _form_class=LoginFor
+                resp.set_cookie(TWO_FACTOR_COOKIE_KEY, token)
+
+                return resp
+            else:
+                # If the user-originating redirection url is not safe, then
+                # redirect to the index instead.
+                if not redirect_to or not is_safe_url(
+                    url=redirect_to, host=request.host
+                ):
+                    redirect_to = request.route_path("manage.projects")
+
+                # Actually perform the login routine for our user.
+                headers = _login_user(request, user_service, userid)
+
+                # Now that we're logged in we'll want to redirect the user to
+                # either where they were trying to go originally, or to the default
+                # view.
+                resp = HTTPSeeOther(redirect_to, headers=dict(headers))
+
+                # We'll use this cookie so that client side javascript can
+                # Determine the actual user ID (not username, user ID). This is
+                # *not* a security sensitive context and it *MUST* not be used
+                # where security matters.
+                #
+                # We'll also hash this value just to avoid leaking the actual User
+                # IDs here, even though it really shouldn't matter.
+                resp.set_cookie(
+                    USER_ID_INSECURE_COOKIE,
+                    hashlib.blake2b(str(userid).encode(), person=b"warehouse.userid")
+                    .hexdigest()
+                    .lower(),
+                )
+
+            return resp
+
+    return {
+        "form": form,
+        "redirect": {"field": REDIRECT_FIELD_NAME, "data": redirect_to},
+    }
+
+
+@view_config(
+    route_name="accounts.two-factor",
+    renderer="accounts/two-factor.html",
+    uses_session=True,
+    require_csrf=True,
+    require_methods=False,
+)
+def two_factor(request, _form_class=TwoFactorForm):
+    if request.authenticated_userid is not None:
+        return HTTPSeeOther(request.route_path("manage.projects"))
+
+    token_service = request.find_service(ITokenService, name="two_factor")
+
+    # Load token data from cookies or
+    # redirect the user to login page again if any token doesn't exist.
+    try:
+        token_data = token_service.loads(request.cookies.get(TWO_FACTOR_COOKIE_KEY))
+    except TokenException:
+        request.session.flash(
+            "Authentication code expired: Try logging in again.", queue="error"
+        )
+        return HTTPSeeOther(request.route_path("accounts.login"))
+
+    userid = token_data.get("userid")
+    if not userid:
+        return HTTPSeeOther(request.route_path("accounts.login"))
+
+    redirect_to = token_data.get("redirect_to")
+
+    user_service = request.find_service(IUserService, context=None)
+
+    form = _form_class(
+        request.POST,
+        user_service=user_service,
+        check_password_metrics_tags=["auth_method:two_factor_form"],
```

Comment:
I think we want these metric tags to match what is on the normal login form: ``["method:auth", "auth_method:login_form"]``.